### ISSUE TITLE
Convention for automatic rechunking

### ISSUE THREAD

**Issue by mrocklin** [dask/dask#3587]
There are a few operations where we can (and possibly should) meaningfully rechunk arrays for users:

1.  atop operations that include outer products, like tensordot
2.  linear algebra operations like qr and svd
3.  fft, where we need active dimensions to have a single chunk
4.  apply_gufunc, where we need core dimensions to have a single chunk
5.  ...

In all of these cases we generally want to make a few dimensions have as few chunks as possible (often one) and then sometimes make other dimensions as large as possible while respecting chunk size memory limits, while still enabling some parallelism.  This can strongly affect performance.

For some operations like `fft` we currently raise informative errors that encourage users to do the right thing.  For other operations like `atop` we don't necessarily want to err, but do want to encourage users to do the right thing.  We might choose to do some of this automatically.  If so, we should do it consistently across all operations in order to avoid API diversity.

I'll propose a `rechunk=` keyword with the following semantics:

1. `rechunk=False` does no rechunking.
2.  `rechunk=True` tells the operation to do whatever it wants to perform automatic rechunking
3.  `rechunk='10MB'` or other string tells the operation to do whatever it wants to perform automatic rechunking, targetting chunk sizes of 10MB.  Otherwise we default to `dask.config.get('array.chunk-size')` as is done now.

I'll also suggest that we leave it to the operations whether or not they choose to do automatic rechunking.

**Comment by mrocklin** [comment]
@jakirkham @shoyer any thoughts or concerns on this topic?

**Comment by shoyer** [comment]
I think the big challenge here is that completely flexible automatic rechunking only makes sense if your dataset can fit into memory without streaming or can be transparently spilled to disk.

My understanding is that this currently works well with the distributed scheduler, but the multi-threaded scheduler could fall flat on its face. We still target the multi-threaded scheduler as a baseline in dask.array, so defaulting to `chunk=True` would potentially be problematic.

Some possible solutions:
- Add spilling to disk to the multi-threaded scheduler.
- `rechunk` only defaults to True if using dask distributed -- maybe set via the config module?

**Comment by mrocklin** [comment]
> I think the big challenge here is that completely flexible automatic rechunking only makes sense if your dataset can fit into memory without streaming or can be transparently spilled to disk.

This is true in the pathological `rechunk((n, 1)).rechunk((1, n))` case but is not true otherwise.  Actually I've found that rechunking before atop operations that have inner and outer products (like tensordot) is often necessary to get good memory performance.

I believe that in all of the operations I mention above, rechunking only helps.

**Comment by mrocklin** [comment]
> but is not true otherwise

Should be 

> but is not universally true, nor necessarily the common case

**Comment by shoyer** [comment]
For something like mean, rechunking along the aggregated dimension could improve performance (probably not enough to pay for rechunking), but would also mean the operation could no longer be done in a streaming way. This would be a bad idea.

For operations like tensordot, you can't really do them in a streaming way across summed dimensions anyways, so it's probably fine to automatically rechunk.

As long as we stick to those sort of operations for `rechunk=True` I think that would be fine.

**Comment by mrocklin** [comment]
Yeah, what I'm hearing is that "we shouldn't automatically rechunk when it might negatively impact streamability".  I agree with that as a restriction.

Do you have any thoughts or concerns about the suggested API?

Do you have any thoughts on whether or not this should be automatic?

**Comment by jakirkham** [comment]
Just to make sure I understand, is the "otherwise..."  statement above meant to be part of item 3 or is it suppose to be item 4?

**Comment by jakirkham** [comment]
Generally this sounds ok (aside from the points @shoyer already raised).

Might include a warning if automatic rechunking is going to be problematic (though we mostly do this anyways). Would just be sure there is a warning for problematic cases anywhere we are automatically rechunking.

On the API point it would be good to understand what saying `'10MB'` means. For example, is this a hard limit or does it have some error bars around it? What is the behavior if it can't quite make this limit?

How do we handle functions that already do some rechunking as it is required (e.g. [`apply_along_axis`]( https://github.com/dask/dask/blob/0.17.5/dask/array/routines.py#L313-L316 ), [`gufunc`s]( https://github.com/dask/dask/blob/946832ff3ec5b48718cc7aad07a556ff554dea9d/dask/array/gufunc.py#L94-L98 ), or others)?

**Comment by shoyer** [comment]
> This is true in the pathological rechunk((n, 1)).rechunk((1, n)) case but is not true otherwise.

Thinking about this again, I think this could potentially be pretty common with `apply_gufunc` (e.g., for a combination of time-series and image based analyses of a movie).

Indeed, we explicitly chose not to rechunk by default in `apply_gufunc` because crashing with out of memory errors is a really bad experience for users -- even though the operation cannot possibly succeed unless it does a rechunk.

In my mind, this leans toward keeping rechunking explicit and opt-in (at least with the threaded scheduler). Unfortunately, dask.array doesn't know which chunks exist merely for efficient parallelism and which chunks are fundamental to how data is stored, so the best choice is probably not to guess at all.

**Comment by mrocklin** [comment]
> In my mind, this leans toward keeping rechunking explicit and opt-in (at least with the threaded scheduler). Unfortunately, dask.array doesn't know which chunks exist merely for efficient parallelism and which chunks are fundamental to how data is stored, so the best choice is probably not to guess at all.

So we can still keep rechunking explicit and opt-in by setting the default value to `False`.  However by having this as a keyword on the function we remove the burden from the user on how best to rechunk.  Either we say "rechunk your data so that all core dimensions have one chunk" and expect them to understand how to do this, or we say "try passing `rechunk=True`".
